WalletWasabi/BitcoinCore/CachedRpcClient.cs

```diff
@@ -0,0 +1,209 @@
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Primitives;
+using NBitcoin;
+using NBitcoin.RPC;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi.BitcoinCore
+{
+	public class CachedRpcClient : RpcClientBase
+	{
+		private CancellationTokenSource _tipChangeCancellationTokenSource;
+
+		public CachedRpcClient(RPCClient rpc, IMemoryCache cache)
+		: base(rpc)
+		{
+			Cache = cache;
+		}
+		
+		public IMemoryCache Cache { get; }
+
+		private CancellationTokenSource TipChangeCancellationTokenSource
+		{
+			get 
+			{
+				if (_tipChangeCancellationTokenSource is null || _tipChangeCancellationTokenSource.IsCancellationRequested)
+				{
+					_tipChangeCancellationTokenSource = new CancellationTokenSource();
+				}
+				return _tipChangeCancellationTokenSource;
+			}
+		}
+
+		public override async Task<uint256> GetBestBlockHashAsync()
+		{
+			string cacheKey = nameof(GetBestBlockHashAsync);
+			if (!Cache.TryGetValue(cacheKey, out uint256 blockHash))
+			{
+				blockHash = await base.GetBestBlockHashAsync().ConfigureAwait(false);
+
+				var cacheEntryOptions = new MemoryCacheEntryOptions()
+					.SetSize(1)
+					// The best hash doesn't change so often so, keep in cache for 4 seconds.
+					.SetAbsoluteExpiration(TimeSpan.FromSeconds(4))
+					.AddExpirationToken(new CancellationChangeToken(TipChangeCancellationTokenSource.Token));
+
+				// Save data in cache.
+				Cache.Set(cacheKey, blockHash, cacheEntryOptions);
+			}
+			return blockHash;
+		}
+
+		public override async Task<Block> GetBlockAsync(uint256 blockHash)
+		{
+			string cacheKey = $"{nameof(GetBlockAsync)}:{blockHash}";
+			if (!Cache.TryGetValue(cacheKey, out Block block))
+			{
+				block = await base.GetBlockAsync(blockHash).ConfigureAwait(false);
+
+				var cacheEntryOptions = new MemoryCacheEntryOptions()
+					.SetSize(10)
+					// There is a block every 10 minutes in average so, keep in cache for 4 seconds.
+					.SetAbsoluteExpiration(TimeSpan.FromSeconds(4));
```

Comment:
Expiration can probably be set much higher right? (Same for height/header)

The problem then is if the entry is not found and we cache the null value, in that case I would indeed set a lower expiration (block found cache for x min, not found cache for y sec)

Reply:
Yes! that's a good idea.

Reply:
Oh wait, the `GetBlockAsync` throws exception if the block is not there, we are not caching anything in that case.